ibis/expr/datatypes/core.py

```diff
@@ -9,68 +9,49 @@
 from numbers import Integral, Real
 from typing import Any, Iterable, Literal, NamedTuple, Optional
 
-import numpy as np
-import pandas as pd
 import toolz
-from multipledispatch import Dispatcher
 from public import public
-from typing_extensions import get_args, get_origin, get_type_hints
+from typing_extensions import Self, get_args, get_origin, get_type_hints
 
 from ibis.common.annotations import attribute
 from ibis.common.collections import FrozenDict, MapSet
+from ibis.common.dispatch import lazy_singledispatch
 from ibis.common.grounds import Concrete, Singleton
 from ibis.common.temporal import IntervalUnit, TimestampUnit
 from ibis.common.validators import Coercible
 
-# TODO(kszucs): we don't support union types yet
-
-dtype = Dispatcher('dtype')
-
-
-@dtype.register(object)
-def dtype_from_object(value, **kwargs) -> DataType:
-    # TODO(kszucs): implement this in a @dtype.register(type) overload once dtype
-    # turned into a singledispatched function because that overload doesn't work
-    # with multipledispatch
-
-    # TODO(kszucs): support Tuple[int, str] and Tuple[int, ...] typehints
-    # in order to support more kinds of typehints follow the implementation of
-    # Validator.from_annotation
-    origin_type = get_origin(value)
-    if origin_type is None:
-        if isinstance(value, type):
-            if issubclass(value, DataType):
-                return value()
-            elif result := _python_dtypes.get(value):
-                return result
-            elif annots := get_type_hints(value):
-                return Struct(toolz.valmap(dtype, annots))
-            elif issubclass(value, bytes):
-                return bytes
-            elif issubclass(value, str):
-                return string
-            elif issubclass(value, Integral):
-                return int64
-            elif issubclass(value, Real):
-                return float64
-            elif value is type(None):
-                return null
-            else:
-                raise TypeError(
-                    f"Cannot construct an ibis datatype from python type `{value!r}`"
-                )
-        else:
-            raise TypeError(
-                f"Cannot construct an ibis datatype from python value `{value!r}`"
-            )
-    elif issubclass(origin_type, Sequence):
-        (value_type,) = map(dtype, get_args(value))
-        return Array(value_type)
-    elif issubclass(origin_type, Mapping):
-        key_type, value_type = map(dtype, get_args(value))
-        return Map(key_type, value_type)
+
+@lazy_singledispatch
+def dtype(value, nullable=True) -> DataType:
+    """Construct an ibis datatype from a python value."""
```

Comment:
```suggestion
	"""Construct an ibis datatype from a python type."""
```
since these "values" are actually types or type descriptors.

Reply:
Updated.